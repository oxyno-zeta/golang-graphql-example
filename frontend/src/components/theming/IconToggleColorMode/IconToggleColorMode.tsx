import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import IconButton, { IconButtonProps } from '@mui/material/IconButton';
import Tooltip from '@mui/material/Tooltip';
import { useTheme } from '@mui/material/styles';
import SvgIcon from '@mui/material/SvgIcon';
import { mdiBrightness4, mdiBrightness7 } from '@mdi/js';
import ColorModeContext from '../../../contexts/ColorModeContext';

interface Props {
  iconButtonProps?: IconButtonProps;
}

const defaultProps = {
  iconButtonProps: {},
};

function IconToggleColorMode({ iconButtonProps }: Props) {
  // Setup translate
  const { t } = useTranslation();
  // Get theme
  const theme = useTheme();
  // Get color mode context
  const colorMode = useContext(ColorModeContext);

  return (
    <Tooltip title={<>{t(theme.palette.mode === 'dark' ? 'common.lightThemeTooltip' : 'common.darkThemeTooltip')}</>}>
      <span>
        <IconButton onClick={colorMode.toggleColorMode} color="inherit" {...iconButtonProps}>
          {theme.palette.mode === 'dark' ? (
            <SvgIcon>
              <path d={mdiBrightness4} />
            </SvgIcon>
          ) : (
            <SvgIcon>
              <path d={mdiBrightness7} />
            </SvgIcon>
          )}
        </IconButton>
      </span>
    </Tooltip>
  );
}

IconToggleColorMode.defaultProps = defaultProps;

export default IconToggleColorMode;
