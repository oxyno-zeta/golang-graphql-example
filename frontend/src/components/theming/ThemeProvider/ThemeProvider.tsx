import React, { ReactNode, useState, useMemo, useContext } from 'react';
import useMediaQuery from '@mui/material/useMediaQuery';
import { ThemeProvider as MuiThemeProvider, createTheme, ThemeOptions } from '@mui/material/styles';
import { PaletteMode } from '@mui/material';
import Cookies from 'universal-cookie';
import ColorModeContext from '../../../contexts/ColorModeContext';
import ConfigContext from '../../../contexts/ConfigContext';

interface Props {
  children: ReactNode;
  themeOptions: ThemeOptions;
}

function ThemeProvider({ children, themeOptions }: Props) {
  // Get config from context
  const cfg = useContext(ConfigContext);
  // Get cookies object
  const cookies = new Cookies();
  // Check prefer color scheme from system
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');
  // Get stored theme mode
  const storedThemeMode = cookies.get('palette-mode');
  // Compute initial value
  let initVal = storedThemeMode;
  if (initVal === null || (initVal !== 'dark' && initVal !== 'light')) {
    initVal = prefersDarkMode ? 'dark' : 'light';
  }

  // State for mode
  const [mode, setMode] = useState<PaletteMode>(initVal as PaletteMode);

  // Expand
  const { configCookieDomain } = cfg;

  // Create color mode context
  const colorMode = useMemo(() => {
    // Set cookie
    const setCookie = (input: PaletteMode) => {
      cookies.set('palette-mode', input, {
        path: '/',
        maxAge: 31536000, // 1 year
        domain: configCookieDomain,
      });
    };

    return {
      toggleColorMode: () => {
        setMode((prevMode) => {
          // Compute new value
          const newVal = prevMode === 'light' ? 'dark' : 'light';
          // Save in storage
          setCookie(newVal);

          return newVal;
        });
      },
      setColorMode: (input: PaletteMode | null) => {
        setMode((prevMode: PaletteMode) => {
          // Compute new value
          const newVal = input || prevMode;
          // Save in storage
          setCookie(newVal);

          return newVal;
        });
      },
    };
  }, []);

  const theme = useMemo(() => {
    // Initialize working copy
    const work = { ...themeOptions };
    // Patch palette mode from options
    // Check if palette object is present
    // Otherwise set it directly
    if (work.palette) {
      work.palette.mode = mode;
    } else {
      work.palette = { mode };
    }

    return createTheme(work);
  }, [mode, themeOptions]);

  return (
    <ColorModeContext.Provider value={colorMode}>
      <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
    </ColorModeContext.Provider>
  );
}

export default ThemeProvider;
