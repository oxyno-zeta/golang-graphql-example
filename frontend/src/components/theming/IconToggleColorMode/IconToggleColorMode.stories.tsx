import React from 'react';
import { StoryFn, Meta } from '@storybook/react-vite';
import CssBaseline from '@mui/material/CssBaseline';
import Typography from '@mui/material/Typography';
import { useTheme } from '@mui/material/styles';
import ThemeProvider from '~components/theming/ThemeProvider';
import IconToggleColorMode, { Props } from './IconToggleColorMode';

export default {
  title: 'Components/theming/IconToggleColorMode',
  component: IconToggleColorMode,
} as Meta<typeof IconToggleColorMode>;

function Info() {
  const theme = useTheme();
  return <Typography>Theme mode:{theme.palette.mode}</Typography>;
}

export const Playbook: StoryFn<typeof IconToggleColorMode> = function C(args: Props) {
  return (
    <ThemeProvider themeOptions={{}}>
      <CssBaseline />
      <IconToggleColorMode {...args} />
      <Info />
    </ThemeProvider>
  );
};
