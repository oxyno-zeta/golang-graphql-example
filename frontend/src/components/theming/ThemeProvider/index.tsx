import ThemeProvider, { Props as ThemeProviderProps } from './ThemeProvider';

export default ThemeProvider;
export type { ThemeProviderProps };
