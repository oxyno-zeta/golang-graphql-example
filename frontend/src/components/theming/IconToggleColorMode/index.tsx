import IconToggleColorMode, { Props as IconToggleColorModeProps } from './IconToggleColorMode';

export default IconToggleColorMode;
export type { IconToggleColorModeProps };
