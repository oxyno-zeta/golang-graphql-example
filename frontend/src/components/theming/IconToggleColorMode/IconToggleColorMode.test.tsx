import React from 'react';
import { render, fireEvent } from '@testing-library/react';
// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { mdiBrightness2, mdiBrightness7 } from '@mdi/js';
import ThemeProvider from '~components/theming/ThemeProvider';

import IconToggleColorMode from './IconToggleColorMode';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

let matchMedia: (q: string) => MediaQueryList;

describe('theming/IconToggleColorMode', () => {
  beforeAll(() => {
    matchMedia = window.matchMedia;
  });

  afterAll(() => {
    window.matchMedia = matchMedia;
  });

  it('should display a light theme icon with light theme and switch to dark', async () => {
    window.matchMedia = jest.fn().mockImplementation((query) => {
      if (query === '(prefers-color-scheme: dark)') {
        return {
          matches: false,
          media: query,
          onchange: null,
          addListener: jest.fn(), // deprecated
          removeListener: jest.fn(), // deprecated
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
          dispatchEvent: jest.fn(),
        };
      }

      return matchMedia(query);
    });

    const { container } = render(
      <ThemeProvider themeOptions={{}}>
        <IconToggleColorMode />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();

    // Find path
    const pathElement = container.querySelector('path');
    expect(pathElement).not.toBeNull();
    expect(pathElement).toHaveAttribute('d', mdiBrightness7);

    // Find button
    const buttonElement = container.querySelector('button');
    expect(buttonElement).not.toBeNull();
    expect(fireEvent.click(buttonElement as Element)).toBeTruthy();

    expect(pathElement).toHaveAttribute('d', mdiBrightness2);

    expect(container).toMatchSnapshot();
  });

  it('should display a dark theme icon with dark theme and switch to light', async () => {
    window.matchMedia = jest.fn().mockImplementation((query) => {
      if (query === '(prefers-color-scheme: dark)') {
        return {
          matches: true,
          media: query,
          onchange: null,
          addListener: jest.fn(), // deprecated
          removeListener: jest.fn(), // deprecated
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
          dispatchEvent: jest.fn(),
        };
      }

      return matchMedia(query);
    });

    const { container } = render(
      <ThemeProvider themeOptions={{}}>
        <IconToggleColorMode />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();

    // Find path
    const pathElement = container.querySelector('path');
    expect(pathElement).not.toBeNull();
    expect(pathElement).toHaveAttribute('d', mdiBrightness2);

    // Find button
    const buttonElement = container.querySelector('button');
    expect(buttonElement).not.toBeNull();
    expect(fireEvent.click(buttonElement as Element)).toBeTruthy();

    expect(pathElement).toHaveAttribute('d', mdiBrightness7);

    expect(container).toMatchSnapshot();
  });
});
