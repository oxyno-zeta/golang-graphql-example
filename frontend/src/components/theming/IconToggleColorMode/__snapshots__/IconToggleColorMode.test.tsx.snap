// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theming/IconToggleColorMode should display a dark theme icon with dark theme and switch to light 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: #fff;
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: inherit;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(255, 255, 255, 0.08);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(255, 255, 255, 0.3);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-2 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <span
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10,2C8.18,2 6.47,2.5 5,3.35C8,5.08 10,8.3 10,12C10,15.7 8,18.92 5,20.65C6.47,21.5 8.18,22 10,22A10,10 0 0,0 20,12A10,10 0 0,0 10,2Z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root emotion-2"
      />
    </button>
  </span>
</div>
`;

exports[`theming/IconToggleColorMode should display a dark theme icon with dark theme and switch to light 2`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: inherit;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-2 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <span
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,8A4,4 0 0,0 8,12A4,4 0 0,0 12,16A4,4 0 0,0 16,12A4,4 0 0,0 12,8M12,18A6,6 0 0,1 6,12A6,6 0 0,1 12,6A6,6 0 0,1 18,12A6,6 0 0,1 12,18M20,8.69V4H15.31L12,0.69L8.69,4H4V8.69L0.69,12L4,15.31V20H8.69L12,23.31L15.31,20H20V15.31L23.31,12L20,8.69Z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root emotion-2"
      />
    </button>
  </span>
</div>
`;

exports[`theming/IconToggleColorMode should display a light theme icon with light theme and switch to dark 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: inherit;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-2 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <span
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,8A4,4 0 0,0 8,12A4,4 0 0,0 12,16A4,4 0 0,0 16,12A4,4 0 0,0 12,8M12,18A6,6 0 0,1 6,12A6,6 0 0,1 12,6A6,6 0 0,1 18,12A6,6 0 0,1 12,18M20,8.69V4H15.31L12,0.69L8.69,4H4V8.69L0.69,12L4,15.31V20H8.69L12,23.31L15.31,20H20V15.31L23.31,12L20,8.69Z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root emotion-2"
      />
    </button>
  </span>
</div>
`;

exports[`theming/IconToggleColorMode should display a light theme icon with light theme and switch to dark 2`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: #fff;
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: inherit;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
}

.emotion-0.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-0 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-0:hover {
  background-color: rgba(255, 255, 255, 0.08);
}

@media (hover: none) {
  .emotion-0:hover {
    background-color: transparent;
  }
}

.emotion-0.Mui-disabled {
  background-color: transparent;
  color: rgba(255, 255, 255, 0.3);
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-2 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <span
    class=""
    data-mui-internal-clone-element="true"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium emotion-0"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10,2C8.18,2 6.47,2.5 5,3.35C8,5.08 10,8.3 10,12C10,15.7 8,18.92 5,20.65C6.47,21.5 8.18,22 10,22A10,10 0 0,0 20,12A10,10 0 0,0 10,2Z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root emotion-2"
      />
    </button>
  </span>
</div>
`;
